lib/Dialect/Arc/Transforms/FindInitialVectors.cpp

```diff
@@ -0,0 +1,222 @@
+  LLVM_DEBUG(llvm::dbgs() << "- Computing topological order in block " << block
+                          << "\n");
+  struct WorklistItem {
+    WorklistItem(Operation *op) : userIt(op->user_begin()) {}
+    Operation::user_iterator userIt;
+    unsigned rank = 0;
+  };
+  SmallMapVector<Operation *, WorklistItem, 16> worklist;
+  for (auto &op : *block) {
+    if (opRanks.contains(&op))
+      continue;
+    worklist.insert({&op, WorklistItem(&op)});
+    while (!worklist.empty()) {
+      auto &[op, item] = worklist.back();
+      if (auto stateOp = dyn_cast<StateOp>(op)) {
+        if (stateOp.getLatency() > 0)
+          item.userIt = op->user_end();
+      } else if (auto writeOp = dyn_cast<MemoryWritePortOp>(op)) {
+        item.userIt = op->user_end();
+      }
+      if (item.userIt == op->user_end()) {
+        opRanks.insert({op, item.rank});
+        worklist.pop_back();
+        continue;
+      }
+      if (auto rankIt = opRanks.find(*item.userIt); rankIt != opRanks.end()) {
+        item.rank = std::max(item.rank, rankIt->second + 1);
+        ++item.userIt;
+        continue;
+      }
+      if (!worklist.insert({*item.userIt, WorklistItem(*item.userIt)}).second)
+        return op->emitError("dependency cycle");
+    }
+  }
+  return success();
+}
+
+namespace {
+template <typename T>
+LogicalResult modifyKeyIfPossible(const T &container, std::string &key) {
+  if(container.empty())
+    return success();
+
+  if constexpr (std::is_same_v<T, StringAttr>) {
+    // The attributes' order doesn't matter so we need to sort the attributes
+    // in some order and the concatinate them
+    std::vector tempAttributeHolder(container.begin(), container.end());
+    std::sort(tempAttributeHolder.begin(), tempAttributeHolder.end());
+    key += std::accumulate(tempAttributeHolder.begin(), tempAttributeHolder.end(), "");
+    return success();
+  }
+
+  // OperandTypes or ResultTypes
+  for (const auto &elem : container) {
+    std::string str;
+    llvm::raw_string_ostream os(str);
+    elem.print(os);
+    key += str;
+  }
+  return success();
+}
+
+std::string computeKey(Operation *op) {
+  // The key = concat(op_rank, op_name, op_operands_types, op_result_types, op_attrs)
+  std::string key = op->getName().getStringRef().str();
+  if (failed(modifyKeyIfPossible(op->getOperandTypes(), key)) ||
+      failed(modifyKeyIfPossible(op->getResultTypes(), key)) || 
+      failed(modifyKeyIfPossible(op->getRegisteredInfo()->getAttributeNames(), key)))
+    assert(false && "Cannot compute the op key");
+
+  return key;
+}
+
+struct Vectorizer {
+  LogicalResult collectSeeds(Block *block) {
+    if (failed(order.compute(block)))
+      return failure();
+
+    for(auto &[op, rank] : order.opRanks)
+      candidates[std::to_string(rank) + computeKey(op)].push_back(op);
+
+    return success();
+  }
+
+  LogicalResult vectorize();
+  // Store Isomorphic ops together
+  std::unordered_map<std::string, SmallVector<Operation *>> candidates;
```

Comment:
You might be able to turn this into a `DenseMap`, which is common in LLVM code bases.